cloudflare.go

```diff
@@ -87,26 +101,72 @@ func (api *API) makeRequest(method, uri string, params interface{}) ([]byte, err
 
 func (api *API) makeRequestWithAuthType(method, uri string, params interface{}, authType int) ([]byte, error) {
 	// Replace nil with a JSON object if needed
-	var reqBody io.Reader
+	var jsonBody []byte
+	var err error
 	if params != nil {
-		json, err := json.Marshal(params)
+		jsonBody, err = json.Marshal(params)
 		if err != nil {
 			return nil, errors.Wrap(err, "error marshalling params to JSON")
 		}
-		reqBody = bytes.NewReader(json)
 	} else {
-		reqBody = nil
+		jsonBody = nil
 	}
 
-	resp, err := api.request(method, uri, reqBody, authType)
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
+	var resp *http.Response
+	var respErr error
+	var reqBody io.Reader
+	var respBody []byte
+	for i := 0; i <= api.retryPolicy.MaxRetries; i++ {
+		if jsonBody != nil {
+			reqBody = bytes.NewReader(jsonBody)
+		}
 
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, errors.Wrap(err, "could not read response body")
+		if i > 0 {
+			// expect the backoff introduced here on errored requests to dominate the effect of rate limiting
+			// dont need a random component here as the rate limiter should do something similar
+			// nb time duration could truncate an arbitrary float. Since our inputs are all ints, we should be ok
+			sleepDuration := time.Duration(math.Pow(2, float64(i-1)) * float64(api.retryPolicy.MinRetryDelay))
+
+			if sleepDuration > api.retryPolicy.MaxRetryDelay {
+				sleepDuration = api.retryPolicy.MaxRetryDelay
+			}
+			// useful to do some simple logging here, maybe introduce levels later
+			log.Printf("Sleeping %s before retry attempt number %d for request %s %s", sleepDuration.String(), i, method, uri)
```

Comment:
Strongly prefer that we didn't log directly in the library.
* Add a simple `Logger` interface to the `API` struct and provide a default no-op method.
* Accept this new `Logger` type as an `Option` - 

```
func Logger(logger *Logger) Option {
	return func(api *API) error {
		api.logger = logger
		return nil
	}
}
```

```
type Logger interface {
	Log(keyvals ...interface{}) error
}
```

Reply:
I like this approach but would suggest aligning with the standard interface so that people can just pass straight in a log.Logger or a logrus.Logger (etc)